### PR TITLE
feat: Read CSV clusters and measurement-particles-map & refactor

### PR DESCRIPTION
**Pull request by benjaminhuth**
**!! This breaks the examples file standards !!**

* Enables reading of cluster and optionally reading of measurement-particles-map in `CsvMeasurmentReader`.
* Adds a roundtrip-test for the Csv measurment reader/writers. It also provides some generic crazy template stuff which hopefully allows to write round-trip tests in a easy way.
* There exists a `hit_id`, which I in the context of the simhit-writer refers to the id of the simulated hit . Additionally to this, there are measurements, which is not a one-to-one mapping. Therefore I would argue we should consistently use the term `measurment_id` when refering to these measurments.

Should now go in after #2330

### CODE REVIEW

**Review by andiwand — APPROVED** [review]

**Comment by benjaminhuth** [comment]
@asalzburger Just pinnging you to make sure I don't break something accidentally here, since this is also changing the file spec. I also haven't updated the `CsvPlanarClusterReader`/`CsvPlanarClusterWriter` (is that infrastructure still needed actually?).

Now it is clear I think that in the EDM we have one cluster per measurement.

I think a next step would be to look again what happens with the clusters in the digitization if we merge measurements.

**Comment by paulgessinger** [comment]
Can you resolve the conflicts @benjaminhuth?

File: Tests/UnitTests/Examples/Io/Csv/MeasurementReaderWriterTests.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    Acts::Vector2 p = Acts::Vector2::Random();
+    Acts::SymMatrix2 c = Acts::SymMatrix2::Random();
+
+    auto m = Acts::makeMeasurement(Acts::SourceLink{sl}, p, c, Acts::eBoundLoc0,
+                                   Acts::eBoundLoc1);
+
+    // TODO this fails!
+    // auto m = Acts::makeMeasurement(Acts::SourceLink{sl}, p, c,
+    // Acts::eBoundLoc0,
+    //                                Acts::eBoundTime);
```
Comment (andiwand) [inline Tests/UnitTests/Examples/Io/Csv/MeasurementReaderWriterTests.cpp:59]: is this still relevant?

**Review by andiwand — DISMISSED** [review]
nice!

File: Tests/UnitTests/Examples/Io/Csv/MeasurementReaderWriterTests.cpp (hunk 1 above)
Comment (benjaminhuth) [reply Tests/UnitTests/Examples/Io/Csv/MeasurementReaderWriterTests.cpp:59]: Hmm yeah it is not supported to write/read non-consecutive measurments currently... I don't plan to implement this at the moment...

File: Examples/Io/Csv/src/CsvMeasurementReader.cpp [hunk 2]
```diff
@@ -112,14 +129,62 @@ std::vector<ActsExamples::MeasurementData> readMeasurementsByGeometryId(
...
+  auto cellDataIt = cellData.begin();
+  for (auto index = 0ul; index < nMeasurments; ++index) {
+    // Search for the range containing the index
+    cellDataIt = std::find_if(cellDataIt, cellData.end(), [&](const auto& cd) {
+      return cd.measurement_id == index;
+    });
+
+    auto nextCellDataIt = std::find_if(
+        cellDataIt, cellData.end(),
+        [&](const auto& cd) { return cd.measurement_id != index; });
```
Comment (andiwand) [inline Examples/Io/Csv/src/CsvMeasurementReader.cpp:149]: would this be a case for `upper_bound` / `lower_bound`? or a multi map?

nothing critical just wondered if the code would be easier to read

**Review by andiwand — APPROVED** [review]
lgtm! just one comment - feel free to ignore

File: Examples/Io/Csv/src/CsvMeasurementReader.cpp (hunk 2 above)
Comment (benjaminhuth) [reply Examples/Io/Csv/src/CsvMeasurementReader.cpp]: good idea, a multimap makes the code a lot more clear actually

**Review by andiwand — DISMISSED** [review]

**Review by andiwand — DISMISSED** [review]

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6055753)!

|status|job|
|------|---|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771936) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771935) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771934) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771933) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771932) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771931) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771930) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771929) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771928) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771927) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771926) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771925) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771924) |
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771923) |
|:red_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771922) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771921) |
|:red_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31771920) |

**Comment by paulgessinger** [comment]
I don't see how this could have interacted with the output tests in Athena.

**Pull request merged**